### PR TITLE
[1st version] Enable tuple and NamedTuple

### PR DESCRIPTION
**Pull request by Mecoli1219**
## Tracking issue
- https://github.com/flyteorg/flyte/issues/5908
- https://github.com/flyteorg/flyte/issues/1337
- https://github.com/flyteorg/flyte/issues/3158
- https://github.com/flyteorg/flyte/issues/4358

## Why are the changes needed?
Check out the RFC

## What changes were proposed in this pull request?
- [x] Run local test
- [x] Run remote test
Currently, the flyte Console doesn't support tuple type on UI, we will open another PR to address that.

#### NamedTuple
```python
import typing
from typing import NamedTuple

class TupleClass(NamedTuple):
    keyi: int
    keys: str

@task(container_image=image)
def nt_t1_1(a1: TupleClass, b1: str) -> typing.Tuple[TupleClass, str]:
    return TupleClass(a1[0], a1[1]), b1

@task(container_image=image)
def nt_t2_1(a2: TupleClass, b2: str) -> typing.Tuple[TupleClass, str]:
    return TupleClass(a2[0], b2), a2[1]

@workflow
def nt_wf1_1(b: str) -> typing.Tuple[TupleClass, str]:
    t, s = nt_t1_1(a1=TupleClass(12345, "NamedTupleData"), b1=b)
    return nt_t2_1(a2=TupleClass(t[0], t[1]), b2=s)

@workflow
def nt_wf2_1(b: str) -> typing.Tuple[TupleClass, str]:
    (at, bt), s = nt_t1_1(a1=TupleClass(12345, "NamedTupleData"), b1=b)
    return nt_t2_1(a2=TupleClass(at, bt), b2=s)
```
#### Tuple
```python
@task
def t_t1(
    a1: typing.Tuple[int, str], b1: str
) -> typing.Tuple[typing.Tuple[int, str], str]:
    return (a1[0], a1[1]), b1

@task
def t_t2(
    a2: typing.Tuple[str, int], b2: str
) -> typing.Tuple[typing.Tuple[int, str], str]:
    return (a2[1], b2), a2[0]

@workflow
def t_wf1(b: str) -> typing.Tuple[typing.Tuple[int, str], str]:
    t, s = t_t1(a1=(54321, "tuple_data"), b1=b)
    return t_t2(a2=(t[1], t[0]), b2=s)

@workflow
def t_wf2(b: str) -> typing.Tuple[typing.Tuple[int, str], str]:
    (at, bt), s = t_t1(a1=(54321, "tuple_data"), b1=b)
    return t_t2(a2=(bt, at), b2=s)
```

### Screenshots
image

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [] All new and existing tests passed.
- [x] All commits are signed-off.

The tests require [IDL](https://github.com/flyteorg/flyte/pull/5720) to be merged.

## Related PRs
- RFC: https://github.com/flyteorg/flyte/pull/5699
- IDL: https://github.com/flyteorg/flyte/pull/5720

### CODE REVIEW

**Comment by kumare3** [comment]
woah!!!!!!!!
@Mecoli1219 you rock!

File: flytekit/core/type_helpers.py [hunk 1]
```diff
@@ -24,3 +24,14 @@ def load_type_from_tag(tag: str) -> typing.Type[T]:
         raise ValueError(f"Could not find the protobuf named: {name} @ {module}.")
 
     return getattr(pb_module, name)
+
+
+def is_namedtuple(t: typing.Type[T]) -> bool:
```
Comment (pingsutw) [inline flytekit/core/type_helpers.py:29]: It would be great to have a unit test for it.

File: flytekit/core/type_engine.py [hunk 2]
```diff
@@ -1419,6 +1428,226 @@ def guess_python_type(self, literal_type: LiteralType) -> list:  # type: ignore
...
+
+    def to_literal(
+        self,
+        ctx: FlyteContext,
+        python_val: T,
+        python_type: Type[T],
+        expected: LiteralType,
+    ) -> Literal:
+        if expected.tuple_type.tuple_name != TupleTransformer.default_tuple_name():
+            # We are expected a NamedTuple
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:1513]: Do you think we need a separate transformer for NamedTuple? `NamedTupleTransformer`

Comment (Mecoli1219) [reply flytekit/core/type_engine.py:1513]: Also, I have tried it before (maybe you can see that in some commits in this Pull Request), and it is really hard to maintain that.

**Pull request closed without merging**